zkp/golang/internal/storage/sql.go

```diff
@@ -0,0 +1,133 @@
+// Copyright © 2024 Kaleido, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"github.com/hyperledger-labs/zeto/internal/node"
+	"github.com/hyperledger-labs/zeto/internal/utxo"
+	"github.com/hyperledger-labs/zeto/pkg/core"
+	"gorm.io/gorm"
+)
+
+const (
+	// we use a table to store the root node indexes for
+	// all the merkle trees in the database
+	TreeRootsTable = "smtRoots"
+	// we use a separate table to store the nodes of each merkle tree
+	// by using the following name as the prefix, followed by the
+	// name of the tree
+	NodesTable_Prefix = "smtNodes_"
+)
+
+type sqlStorage struct {
+	p              core.SqlDBProvider
+	smtName        string
+	nodesTableName string
+}
+
+// NewSqlStorage creates a new sqlStorage instance
+// The "smtName" is the name for the tree instance, it must
+// be unique within the backend database instance
+func NewSqlStorage(p core.SqlDBProvider, smtName string) *sqlStorage {
+	return &sqlStorage{
+		p:              p,
+		smtName:        smtName,
+		nodesTableName: NodesTable_Prefix + smtName,
+	}
+}
+
+func (s *sqlStorage) GetRootNodeIndex() (core.NodeIndex, error) {
+	root := core.SMTRoot{
+		Name: s.smtName,
+	}
+	err := s.p.DB().Table(TreeRootsTable).First(&root).Error
+	if err == gorm.ErrRecordNotFound {
+		return nil, ErrNotFound
+	} else if err != nil {
+		return nil, err
+	}
+	idx, err := node.NewNodeIndexFromHex(root.RootIndex)
+	return idx, err
+}
+
+func (m *sqlStorage) UpsertRootNodeIndex(root core.NodeIndex) error {
+	err := m.p.DB().Table(TreeRootsTable).Save(&core.SMTRoot{
+		RootIndex: root.Hex(),
+		Name:      m.smtName,
+	}).Error
+	return err
+}
+
+func (m *sqlStorage) GetNode(ref core.NodeIndex) (core.Node, error) {
+	// the node's reference key (not the index) is used as the key to
+	// store the node in the DB
+	n := core.SMTNode{
+		RefKey: ref.Hex(),
+	}
+	err := m.p.DB().Table(m.nodesTableName).First(&n).Error
```

Comment:
I wonder what it safer:
- To have a Golang RLock?
- To have an actual DB lock on write?

Reply:
I think in our case implementing the locks in golang might be the way to do, as there are complex call paths that goes from writing to reading, such as when `addLeaf()` calls `getNode()`. it'd be very complex and dead-lock prone to try to do the locks inside the DB calls in the storage package.